### PR TITLE
[corlib] Default to fast Span<T>

### PR DESCRIPTION
**Pull request by marek-safar**
Part of #7249

### CODE REVIEW

**Comment by vargaz** [comment]
This needs a rebase.

**Comment by vargaz** [comment]
Seems to be missing Buffer:Memmove<T>, also the definition of the non generic version of memmove doesn't seem right:

                internal static unsafe void Memmove (byte *dest, byte *src, uint len)
                {
                        Memcpy (dest, src, (int) len);
                }

                internal static void Memmove<T>(ref T destination, ref T source, nuint elementCount)
                {
                        throw new NotImplementedException ("Buffer::Memmove<T>");
                }

**Comment by marek-safar** [comment]
@vargaz correct, not sure if we should do ical or try to adopt quite ugly managed implementation from coreclr

**Comment by vargaz** [comment]
The build doesn't define BIT64, and some code like Span.Clear () depends on it being the correct value:

               SpanHelpers.ClearWithReferences(ref Unsafe.As<T, IntPtr>(ref _pointer.Value), (nuint)_length * (nuint)(Unsafe.SizeOf<T>() / sizeof(nuint)));

**Comment by vargaz** [comment]
The build seems to work now but there are a lot of class lib failures.

File: mcs/class/System/System.Net/HttpWebRequest.cs [hunk 1]
```diff
@@ -828,7 +828,7 @@ void AddRange (string rangeSpecifier, long from, long to)
 		WebOperation SendRequest (bool redirecting, BufferOffsetSize writeBuffer, CancellationToken cancellationToken)
 		{
 			lock (locker) {
-				WebConnection.Debug ($"HWR SEND REQUEST: Req={ID} requestSent={requestSent} redirecting={redirecting}");
+				WebConnection.Debug ($"HWR SEND REQUEST: Req={ID} requestSent={requestSent} actualUri={actualUri} redirecting={redirecting}");
```
Comment (kg) [inline mcs/class/System/System.Net/HttpWebRequest.cs]: Minor issue but won't $"..." generate garbage and burn cycles string formatting even if debug messages are suppressed? Wouldn't it be better to do "... {0} ...", x style instead

Comment (akoeplinger) [reply mcs/class/System/System.Net/HttpWebRequest.cs]: In general yes but `WebConnection.Debug()` is marked as `[Conditional ("MONO_WEB_DEBUG")]` so the compiler will not even emit it into the output assembly.

**Comment by vargaz** [comment]
The 32 bit failures are (mostly) caused by the BIT64 define when compiling corlib, some coreclr code depends on a correct type for nuint:

       public void Clear()
        {
            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
            {
                SpanHelpers.ClearWithReferences(ref Unsafe.As<T, IntPtr>(ref _pointer.Value), (nuint)_length * (nuint)(Unsafe.SizeOf<T>() / sizeof(nuint)));
            }
            else
            {
                SpanHelpers.ClearWithoutReferences(ref Unsafe.As<T, byte>(ref _pointer.Value), (nuint)_length * (nuint)Unsafe.SizeOf<T>());
            }
        }

**Comment by marek-safar** [comment]
@monojenkins commit apidiff and csproj

**Comment by marek-safar** [comment]
@monojenkins backport 2018-06

**Pull request merged**